lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -1,94 +1,477 @@
-"""
-Helper functions to get the elasticsearch specific configurations
-"""
-
+import requests
+from datetime import datetime
+from dateutil import parser as date_parser
 from dateutil import rrule
+from dateutil.relativedelta import relativedelta
+from enum import Enum
+from flask import request
+from flask_restful import Resource, abort
+from logging import Logger
+from typing import Any, AnyStr, Callable, Dict, List
+from urllib.parse import urljoin
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.auth import Auth
 
-from configparser import NoSectionError, NoOptionError
 
+class SchemaError(TypeError):
+    """
+    Generic base class for errors in processing a JSON schema.
+    """
 
-def get_user_term(user: str):
-    """Elasticsearch doesn't index (or search for) null values, and it's not
-    clear we really want to consider "owner": null to be "owned" datasets, in
-    any case. (Nor will they be possible beyond the transition period where we
-    may retain limited support for 0.69 data.)
+    pass
 
-    Instead, let a query for data with a missing or null "user" parameter find
-    all data that's set to allow public access. This by default includes all
-    "unowned" data, but also data owned by a user that's been published.
+
+class InvalidRequestPayload(SchemaError):
+    """
+    A required client JSON input document is missing.
     """
-    if user:
-        term = {"authorization.owner": user}
-    else:
-        term = {"authorization.access": "public"}
-    return term
 
+    def __str__(self):
+        return "Invalid request payload"
 
-def get_es_host(config):
-    try:
-        return config.get("elasticsearch", "host")
-    except (NoOptionError, NoSectionError):
-        return ""
 
+class MissingParameters(SchemaError):
+    """
+    One or more required JSON keys are missing, or the values are unexpectedly
+    empty.
+    """
 
-def get_es_port(config):
-    try:
-        return config.get("elasticsearch", "port")
-    except (NoOptionError, NoSectionError):
-        return ""
+    def __init__(self, keys: List[AnyStr]):
+        self.keys = keys
+
+    def __str__(self):
+        return f"Missing required parameters: {','.join(self.keys)}"
+
+
+class ConversionError(SchemaError):
+    """
+    ConversionError Used to report an invalid parameter type
+    """
+
+    def __init__(self, value: Any, expected_type: str, actual_type: str):
```

Comment:
That's mostly due to the Username conversion, which is really str -> str, but with a special assumed context. The argument is "username" rather than a type.

I could have reduced some of this fluff though, even so, by having the exception do something like `x = expected_type.__name__ if isinstance(expected_type, type) else expected_type` to avoid the less readable `<class 'name'>` output without requiring all the callers to do it.

On the other hand, that means dropping back to `Any` for the type hint, which isn't very useful.

Reply:
It occurs to me that the mid-ground option is available:
```
    def __init__(self, value: Any, expected_type: str, actual_type: [str, None] = None):
        if actual_type is None:
            actual_type = type(value).__name__
```
that way, you can specify it if you want to, and it defaults naturally if you don't.

For the record, I'm in favor of using common code to reduce fluff and avoid less readable output.  :-)

Reply:
Yes, I could combine both ideas and allow defaulting the actual type _and_ specifying a type or a type name; not sure that's really worthwhile at this point, but I might noodle with this some more later after this conversation. ;-)